### PR TITLE
Alerting: Fix incorrect decoding for alert rules with % characters

### PR DESCRIPTION
**Pull request by gillesdemey**
**What is this feature?**

This PR fixes a crash in the front-end when trying to view the details for an alert rule that has a "%" character encoded in the alert rule name.

This can be traced to a bug in react-router (https://github.com/remix-run/history/issues/505).

the rule editor was also affected, it didn't crash but also did not properly decode the rule identifier.

**Special notes for your reviewer:**

Some additional manual testing would be appreciated since I failed to successfully write a test to mock `window.location`

### CODE REVIEW

**Comment by gillesdemey** [comment]
Tests are currently failing because they rely pretty heavily on mocked location data passed in to the ruler viewer – I'll need to find a way around that to unblock this PR.

**Review by soniaAguilarPeiron — APPROVED** [review]
LGTM!
Tested on my local ✅

**Review by ashharrison90 — COMMENTED** [review]
- are we confident this patch to `history` won't cause unintended behaviour elsewhere in the app?
- if so, can we please create an issue to upgrade `react-router` + `history`, including the context from https://github.com/grafana/grafana/pull/76148/files#diff-d39aa8bb5f5c32177ba0cef60b1277b58288013d5d2ac2daaa68930771e4c350R251-R256, and also reminding us to remove the patch. i don't know where that issue should live - tbh it probably ends up on frontend platform's board 😬

File: public/app/features/alerting/unified/utils/rule-id.ts [hunk 1]
```diff
@@ -244,3 +247,17 @@ function hashLabelsOrAnnotations(item: Labels | Annotations | undefined): string
...
+
+// DO NOT USE REACT-ROUTER HOOKS FOR THIS CODE
+// React-router's useLocation/useParams/props.match are broken and don't preserve original param values when parsing location
+// so, they cannot be used to parse name and sourceName path params
+// React-router messes the pathname up resulting in a string that is neither encoded nor decoded
+// Relevant issue: https://github.com/remix-run/history/issues/505#issuecomment-453175833
+// It was probably fixed in React-Router v6
+type PathWithOptionalID = { id?: string };
+export function getRuleIdFromPathname(params: PathWithOptionalID): string | undefined {
+  const { pathname = '' } = locationService.getLocation();
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/utils/rule-id.ts:259]: Now that we're patching the router, do we still need this workaround? Can we go back to using `useParams()`?

Comment (konrad147) [reply public/app/features/alerting/unified/utils/rule-id.ts:259]: I need to verify. IIRC `react-router-dom` has the same bug. In this PR we only patch the `history` package so it doesn't break outgoing links, but it shouldn't affect parsing path params

Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/utils/rule-id.ts:259]: It seems that react-router added a fix in [here](https://github.com/remix-run/react-router/pull/9477) , and this fix was included in the [v6.4.3](https://github.com/remix-run/react-router/releases/tag/react-router%406.4.3) release

**Comment by soniaAguilarPeiron** [comment]
> * are we confident this patch to `history` won't cause unintended behaviour elsewhere in the app?
> * if so, can we please create an issue to upgrade `react-router` + `history`, including the context from https://github.com/grafana/grafana/pull/76148/files#diff-d39aa8bb5f5c32177ba0cef60b1277b58288013d5d2ac2daaa68930771e4c350R251-R256, and also reminding us to remove the patch. i don't know where that issue should live - tbh it probably ends up on frontend platform's board 😬

I created this [issue](https://github.com/grafana/grafana/issues/76744) @ashharrison90 , should I assign to the Grafana Frontend platform project?

**Comment by ashharrison90** [comment]
@soniaAguilarPeiron sure 👍

**Comment by soniaAguilarPeiron** [comment]
> * are we confident this patch to `history` won't cause unintended behaviour elsewhere in the app?

@ashharrison90  This patch does the same as this [fix](https://github.com/remix-run/history/pull/656) , that has been merged to main. So, we believe it should not break anything. We tested locally and on the ephemeral instance. We cannot be 100% confident about not breaking anything, but we believe it should not.

**Review by konrad147 — APPROVED** [review]
LGTM!

**Pull request merged**